wpilibj/src/main/java/edu/wpi/first/wpilibj/experimental/command/CommandScheduler.java

```diff
@@ -0,0 +1,465 @@
+      }
+    } else {
+      //Else check if the requirements that are in use have all have interruptible commands,
+      //and if so, interrupt those commands and schedule the new command.
+      boolean allInterruptible = true;
+      for (Subsystem requirement : requirements) {
+        if (m_requirements.keySet().contains(requirement)) {
+          allInterruptible &=
+              m_scheduledCommands.get(m_requirements.get(requirement)).isInterruptible();
+        }
+      }
+      if (allInterruptible) {
+        for (Subsystem requirement : requirements) {
+          if (m_requirements.containsKey(requirement)) {
+            cancelCommand(m_requirements.get(requirement));
+          }
+        }
+        command.initialize();
+        CommandState scheduledCommand = new CommandState(interruptible);
+        m_scheduledCommands.put(command, scheduledCommand);
+        for (Consumer<Command> action : m_initActions) {
+          action.accept(command);
+        }
+      }
+    }
+  }
+
+  /**
+   * Schedules multiple commands for execution.  Does nothing if the command is already scheduled.
+   * If a command's requirements are not available, it will only be started if all the commands
+   * currently using those requirements have been scheduled as interruptible.  If this is
+   * the case, they will be interrupted and the command will be scheduled.
+   *
+   * @param interruptible whether the commands should be interruptible
+   * @param commands      the commands to schedule
+   */
+  public void scheduleCommands(boolean interruptible, Command... commands) {
+    for (Command command : commands) {
+      scheduleCommand(command, interruptible);
+    }
+  }
+
+  /**
+   * Runs a single iteration of the scheduler.  The execution occurs in the following order:
+   *
+   * <p>Subsystem periodic methods are called.
+   *
+   * <p>Button bindings are polled, and new commands are scheduled from them.
+   *
+   * <p>Currently-scheduled commands are executed.
+   *
+   * <p>End conditions are checked on currently-scheduled commands, and commands that are finished
+   * have their end methods called and are removed.
+   *
+   * <p>Any subsystems not being used as requirements have their default methods started.
+   */
+  @SuppressWarnings({"PMD.CyclomaticComplexity", "PMD.NPathComplexity"})
+  public void run() {
+
+    if (m_disabled) {
+      return;
+    }
+
+    //Run the periodic method of all registered subsystems.
+    for (Subsystem subsystem : m_subsystems.keySet()) {
+      subsystem.periodic();
+    }
+
+    //Poll buttons for new commands to add.
+    for (ButtonScheduler button : m_buttons) {
+      button.execute();
+    }
+
+    //Run scheduled commands, remove finished commands.
+    for (Iterator<Command> iterator = m_scheduledCommands.keySet().iterator();
+         iterator.hasNext(); ) {
+      Command command = iterator.next();
+
+      if (RobotState.isDisabled() && !command.runsWhenDisabled()) {
```

Comment:
I can't remember but I think there was some bug where checking the robot state could be slower than expected.  Maybe we can flip the order of the operands so the known boolean check will run first and might short circuit.

@ThadHouse was that bug fixed?

Reply:
It wasn't IsDisabled, it was checking for enabled teleop or enabled auto, and it was really only a pythong bug. But flipping the order might be nice because runsWhenDisabled likely doesn't take a lock, whereas isDisabled does.